static/js/copy_and_paste.js

```diff
@@ -208,7 +208,11 @@ exports.paste_handler_converter = function (paste_html) {
     var div = document.createElement("div");
     div.innerHTML = markdown_html;
     // Using textContent for modern browsers, innerText works for Internet Explorer
-    return div.textContent || div.innerText || "";
+    var markdown_text = div.textContent || div.innerText || "";
+    markdown_text = markdown_text.trim();
+    // Removes newlines before the start of a list and between list elements.
+    markdown_text = markdown_text.replace(/\n+([*+-])/g, '\n$1');
```

Comment:
I wonder whether this is the most robust strategy, or we should instead just be removing double-newlines?

Reply:
So this will deal with the newline just before a list starts, as well as some extra newlines between the list elements which you're copying from rich text. When I copied a bulleted list from Dropbox Paper, the HTML was this

```
<div class="ace-line gutter-author-d-iz88z86z86za0dz67zz78zz78zz74zz68zjz80zz71z9iz90za3z66zs0z65zz65zq8z75zlaz81zcz66zj6g2mz78zz76zmz66z22z75zfcz69zz66z ace-ltr focused-line" dir="auto" id="editor-1-ace-line-30"><span>To lower the stress of the interview:</span></div><div class="ace-line gutter-author-d-iz88z86z86za0dz67zz78zz78zz74zz68zjz80zz71z9iz90za3z66zs0z65zz65zq8z75zlaz81zcz66zj6g2mz78zz76zmz66z22z75zfcz69zz66z line-list-type-bullet ace-ltr" dir="auto" id="editor-1-ace-line-31"><ul class="listtype-bullet listindent1 list-bullet1"><li><span class="ace-line-pocket-zws" data-faketext="" data-contentcollector-ignore-space-at="end">​</span><span class="ace-line-pocket" data-faketext="" contenteditable="false"></span><span class="ace-line-pocket-zws" data-faketext="" data-contentcollector-ignore-space-at="start">​</span><span>have all the relevant documents printed</span><span class=" s-lparen"> </span><span class=" h-lparen">(keep</span><span> them all for later!):</span></li></ul></div><div class="ace-line gutter-author-d-iz88z86z86za0dz67zz78zz78zz74zz68zjz80zz71z9iz90za3z66zs0z65zz65zq8z75zlaz81zcz66zj6g2mz78zz76zmz66z22z75zfcz69zz66z line-list-type-bullet ace-ltr" dir="auto" id="editor-1-ace-line-32"><ul class="listtype-bullet listindent2 list-bullet2"><li><span class="ace-line-pocket-zws" data-faketext="" data-contentcollector-ignore-space-at="end">​</span><span class="ace-line-pocket" data-faketext="" contenteditable="false"></span><span class="ace-line-pocket-zws" data-faketext="" data-contentcollector-ignore-space-at="start">​</span><span>information proving you’re a programmer</span><span class=" s-lparen"> </span><span class=" h-lparen">(especially</span><span>
 if you’re a woman and/or look young), in a form understandable to 
non-programmers - I’ve used my Recurse Center acceptance email and info 
from their website, you could use info from Zulip website and some info 
on the GCI</span><span class=" s-lparen"> </span><span class=" h-lparen">(Google</span><span> logo could help a lot during the interview)</span></li></ul></div>
```

and the original copy-paste output was

```
To lower the stress of the interview:



*   ​​have all the relevant documents printed  (keep them all for later!):





*   ​​information proving you’re a programmer  (especially if you’re a woman and/or look young), in a form understandable to non-programmers - I’ve used my Recurse Center acceptance email and info from their website, you could use info from Zulip website and some info on the GCI  (Google logo could help a lot during the interview)
```

Reply:
Cool, makes sense.  The other test case that I tried just now was a paper bulleted list with some nesting in the bullets; that seems to not do the right thing (not sure if that's a Paper bug or our bug, though).    



I think we should collect a few more test cases like

Reply:
Yeah, I believe toMarkdown doesn't work well with nested bullets, it just prints that as one single list. 

I added a testcase from Github and Dropbox, will try some more.

Reply:
Cool, for nested bullets, maybe we should submit an issue to upstream toMarkdown so that they're tracking that as a bug in their tool?